### PR TITLE
[ENG-2108][ENG-2107] Purge archive fails

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Expands on PR purging files for withdrawn registrations to include failed registrations as well

## Changes

- adds periodic tasks to mark files as deleted and purge them, with tests.

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify failed registrations are purged, check with dev-ops to ensure files are purged properly.

## Documentation

new tech task, no docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-2108

### CODE REVIEW

File: osf/models/registrations.py [hunk 1]
```diff
@@ -123,11 +125,15 @@ class Registration(AbstractNode):
...
-    def find_failed_registrations():
+    def find_failed_registrations(include_deleted=False):
         expired_if_before = timezone.now() - settings.ARCHIVE_TIMEOUT_TIMEDELTA
         node_id_list = ArchiveJob.objects.filter(sent=False, datetime_initiated__lt=expired_if_before, status=ARCHIVER_INITIATED).values_list('dst_node', flat=True)
         root_nodes_id = AbstractNode.objects.filter(id__in=node_id_list).values_list('root', flat=True).distinct()
-        stuck_regs = AbstractNode.objects.filter(id__in=root_nodes_id, is_deleted=False)
+        stuck_regs = AbstractNode.objects.filter(id__in=root_nodes_id)
+
+        if not include_deleted:
+            stuck_regs = stuck_regs.filter(is_deleted=False)
```
Comment (corbinSanders) [inline osf/models/registrations.py:135]: **Eventually** we will get rid of `is_deleted` in favor of `deleted`.

File: website/settings/defaults.py [hunk 2]
```diff
@@ -643,6 +643,13 @@ class CeleryConfig:
                 'task': 'management.commands.update_institution_project_counts',
                 'schedule': crontab(minute=0, hour=9), # Daily 05:00 a.m. EDT
             },
+            'purge_files': {
+                'task': 'management.commands.purge_withdrawn_files',
```
Comment (corbinSanders) [inline website/settings/defaults.py:647]: You'll need to add this management command to the Celery Config above as well.

File: osf/management/commands/purge_files.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
+import logging
+
+from django.utils import timezone
+
+from framework.celery_tasks import app as celery_app
+from website.app import setup_django
+setup_django()
```
Comment (corbinSanders) [inline osf/management/commands/purge_files.py:7]: Nit: Could you move this to line 18.

File: osf/management/commands/purge_files.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+def purge_deleted_failed_registration_files(dry_run=False):
+    deleted_file_ids_on_failed_registration = Registration.find_failed_registrations(include_deleted=True).filter(
+        files__deleted__lte=timezone.now() - PURGE_DELTA
+    ).values_list('files__id', flat=True).distinct()
+
+    creds = Credentials.from_service_account_file(GCS_CREDS)
+    client = Client(credentials=creds)
+    for file in TrashedFileNode.objects.filter(id__in=deleted_file_ids_on_failed_registration):
+        if not dry_run:
+            file._purge(client=client)
```
Comment (corbinSanders) [inline osf/management/commands/purge_files.py:62]: I **think** you can just wrap this in a `with transaction.atomic()` and error out of it for a dry run.

File: osf/models/registrations.py [hunk 5]
```diff
@@ -681,6 +684,42 @@ def remove_tags(self, tags, auth, save=True):
         else:
             raise NodeStateError('Cannot remove tags of withdrawn registrations.')
 
+    def withdraw(self):
```
Comment (corbinSanders) [inline osf/models/registrations.py:687]: Should there be a registration log action for withdrawn?

**Review by corbinSanders — CHANGES_REQUESTED** [review]
Looks really good! A few comments (see below)...

**Comment by corbinSanders** [comment]
LGTM! Thanks!

**Comment by Johnetordoff** [comment]
@jwalz One thing that we talked about during final review was actually wrong, we aren't repeatedly deleting things already marked deleted because the `.files` relationship on the node only is for `OsfStorageFile`s and not `TrashedFileNode`s so things marked and recast as deleted are filtered by default.

File: addons/osfstorage/models.py [hunk 6]
```diff
@@ -164,7 +164,7 @@ def is_preprint_primary(self):
     def delete(self, user=None, parent=None, **kwargs):
```
Comment (mfraezz) [inline addons/osfstorage/models.py:164]: Should `parent` be removed here as well?

File: addons/osfstorage/models.py [hunk 7]
```diff
@@ -387,7 +387,7 @@ def delete(self, user=None, parent=None, **kwargs):
         from website.search import search
 
         search.update_file(self, delete=True)
-        return super(OsfStorageFile, self).delete(user, parent, **kwargs)
+        return super().delete(user, parent, **kwargs)
```
Comment (mfraezz) [inline addons/osfstorage/models.py:390]: `parent` must be removed here as well (and preferably from the function definition too) -- `super`'s call won't have it, so its default of `None` overrides `save=True` as the next positional argument in `BaseFileNode.delete`.

Also, not your `git blame`, but prefer passing keyword arguments with their keywords to avoid this type of issue.

File: osf/models/files.py [hunk 8]
```diff
@@ -410,13 +410,11 @@ def _update_node(self, recursive=True, save=True):
             for child in self.children:
                 child._update_node(save=save)
 
-    # TODO: Remove unused parent param
-    def delete(self, user=None, parent=None, save=True, deleted_on=None):
+    def delete(self, user=None, save=True, deleted_on=None):
```
Comment (mfraezz) [inline osf/models/files.py:413]: `TrashedFileNode` in this file also has a `parent=None` to remove

File: osf_tests/fakegcscreds.pem [hunk 9]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (mfraezz) [inline osf_tests/fakegcscreds.pem:1]: This probably doesn't need to exist.

File: osf_tests/test_elastic_search.py [hunk 10]
```diff
@@ -811,7 +811,7 @@ def test_retraction_wiki_content_is_not_searchable(self):
         self.registration.retraction.state = Retraction.APPROVED
         with run_celery_tasks():
             self.registration.retraction.save()
-            self.registration.save()
```
Comment (mfraezz) [inline osf_tests/test_elastic_search.py:814]: Minor: Why was this removed?

File: website/settings/defaults.py [hunk 11]
```diff
@@ -408,6 +408,7 @@ class CeleryConfig:
         'scripts.analytics.run_keen_snapshots',
         'scripts.analytics.run_keen_events',
         'scripts.clear_sessions',
+        'osf.management.commands.purge_files',
```
Comment (mfraezz) [inline website/settings/defaults.py:411]: This name, and the file itself, should probably be changed to something more like `delete_withdrawn_or_failed_registration_files`

File: osf_tests/factories.py [hunk 12]
```diff
@@ -453,20 +453,20 @@ def add_approval_step(reg):
...
 
-        registration = kwargs.pop('registration', None)
+        registration = kwargs.pop('registration', RegistrationFactory())
         registration.is_public = True
         user = kwargs.pop('user', registration.creator)
 
         registration.retract_registration(user)
         withdrawal = registration.retraction
         token = list(withdrawal.approval_state.values())[0]['approval_token']
-        with patch('osf.models.AbstractNode.update_search'):
```
Comment (mfraezz) [inline osf_tests/factories.py:468]: Minor: why was this context removed?

File: website/settings/defaults.py [hunk 13]
```diff
@@ -645,6 +646,13 @@ class CeleryConfig:
                 'task': 'management.commands.update_institution_project_counts',
                 'schedule': crontab(minute=0, hour=9), # Daily 05:00 a.m. EDT
             },
+            'purge_files': {
+                'task': 'management.commands.purge_withdrawn_files',
```
Comment (mfraezz) [inline website/settings/defaults.py:650]: Names don't match, the decorator and import are both `[...].purge_files`

File: website/settings/defaults.py [hunk 14]
```diff
@@ -645,6 +646,13 @@ class CeleryConfig:
...
+                'schedule': crontab(minute=0, hour=5),  # Daily 12 a.m
+                'kwargs': {'dry_run': False}
+
+            },
+
```
Comment (mfraezz) [inline website/settings/defaults.py:655]: Minor: inconsistent whitespace in this dict.

File: website/settings/local-travis.py [hunk 15]
```diff
@@ -106,4 +106,6 @@ class CeleryConfig(defaults.CeleryConfig):
 DOI_FORMAT = '{prefix}/FK2osf.io/{guid}'
 
 SHARE_ENABLED = False
+
+GCS_CREDS = 'osf_tests/fakegcscreds.pem'
```
Comment (mfraezz) [inline website/settings/local-travis.py:110]: This can be removed.

File: website/settings/defaults.py [hunk 16]
```diff
@@ -2023,3 +2031,6 @@ def from_node_usage(cls,  usage_bytes, private_limit=None, public_limit=None):
             return cls.DEFAULT
 
 STORAGE_USAGE_CACHE_TIMEOUT = 3600 * 24  # seconds in hour times hour (one day)
+
+# Google Cloud storage credentials, our default osfstorage backend
+GCS_CREDS = None
```
Comment (mfraezz) [inline website/settings/defaults.py:2036]: This can be removed.

**Review by mfraezz — CHANGES_REQUESTED** [review]
Pass complete. Solid core, but needs to do less rather than more. :octocat:

File: osf_tests/factories.py (hunk 12 above)
Comment (Johnetordoff) [reply osf_tests/factories.py:468]: It had no effect just superfluous mocking I removed for debugging, it didn't effect the test so I left it out.

Comment (mfraezz) [reply osf_tests/factories.py:468]: Seemingly, it prevented a potentially-unnecessary call to ES, i.e. sped up runtime slightly for every call to `WithdrawnRegistrationFactory([...])`

File: osf/management/commands/delete_withdrawn_or_failed_registration_files.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    failed_registrations = Registration.find_failed_registrations(days_stuck=STUCK_FILES_DELETE_TIMEOUT)
+    logger.info(f'{"[DRY-RUN]" if dry_run else ""} There are {failed_registrations.count()} registrations that are stuck having there files marked as deleted')
+
+    for reg in failed_registrations.annotate(fc=Count('files')).filter(fc__gte=1)[:batch_size]:
+        reg.delete_registration_tree(save=True)
+        files_to_be_deleted = reg.files.all()
+        logger.info(f'{"[DRY-RUN]" if dry_run else ""} There are {files_to_be_deleted.count()} files deleted from stuck registration ({reg._id})')
+        for file in files_to_be_deleted:
+                file.delete()
```
Comment (mfraezz) [inline osf/management/commands/delete_withdrawn_or_failed_registration_files.py:41]:
```suggestion
            if not dry_run:
                file.delete()
```

File: osf/models/registrations.py [hunk 18]
```diff
@@ -128,11 +130,15 @@ class Registration(AbstractNode):
...
-    def find_failed_registrations():
-        expired_if_before = timezone.now() - settings.ARCHIVE_TIMEOUT_TIMEDELTA
+    def find_failed_registrations(include_deleted=False, days_stuck=None):
+        expired_if_before = timezone.now() - (days_stuck or settings.ARCHIVE_TIMEOUT_TIMEDELTA)
         node_id_list = ArchiveJob.objects.filter(sent=False, datetime_initiated__lt=expired_if_before, status=ARCHIVER_INITIATED).values_list('dst_node', flat=True)
         root_nodes_id = AbstractNode.objects.filter(id__in=node_id_list).values_list('root', flat=True).distinct()
-        stuck_regs = AbstractNode.objects.filter(id__in=root_nodes_id, is_deleted=False)
+        stuck_regs = AbstractNode.objects.filter(id__in=root_nodes_id)
+
+        if not include_deleted:
```
Comment (mfraezz) [inline osf/models/registrations.py:139]: This block / kwarg is no longer used, can it be reverted?

File: osf/management/commands/delete_withdrawn_or_failed_registration_files.py (hunk 17 above)
Comment (Johnetordoff) [reply osf/management/commands/delete_withdrawn_or_failed_registration_files.py:41]: good catch

File: osf/management/commands/delete_withdrawn_or_failed_registration_files.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    This script purges files that are deleted after being withdrawn 30 days.
+    """
+    if dry_run:
+        logger.info('This is a dry run; no files will be purged or marked as deleted.')
+
+    # Withdrawn files
+    mark_withdrawn_files_as_deleted(dry_run, batch_size_withdrawn)
+
+    # Failed Registration files
+    mark_failed_registration_files_as_deleted(dry_run, batch_size_stuck)
```
Comment (mfraezz) [inline osf/management/commands/delete_withdrawn_or_failed_registration_files.py:57]:
```suggestion
    mark_failed_registration_files_as_deleted(dry_run, batch_size_stuck)
    if dry_run:
        raise RuntimeError('Dry run, transaction rolled back.')
```
That `reg.delete_registration_tree(save=True)` call above would persist, otherwise.

File: website/settings/local-travis.py [hunk 20]
```diff
@@ -106,4 +106,5 @@ class CeleryConfig(defaults.CeleryConfig):
 DOI_FORMAT = '{prefix}/FK2osf.io/{guid}'
 
 SHARE_ENABLED = False
+
```
Comment (mfraezz) [inline website/settings/local-travis.py:109]: Nitpick, but extra whitespace that wasn't there before in an otherwise unchanged file.

**Review by mfraezz — APPROVED** [review]
Left a couple minor 1-off comments, otherwise LGTM 🍏

File: osf/models/registrations.py [hunk 21]
```diff
@@ -128,11 +130,12 @@ class Registration(AbstractNode):
...
 
     @staticmethod
-    def find_failed_registrations():
-        expired_if_before = timezone.now() - settings.ARCHIVE_TIMEOUT_TIMEDELTA
+    def find_failed_registrations(days_stuck=None):
+        expired_if_before = timezone.now() - (days_stuck or settings.ARCHIVE_TIMEOUT_TIMEDELTA)
         node_id_list = ArchiveJob.objects.filter(sent=False, datetime_initiated__lt=expired_if_before, status=ARCHIVER_INITIATED).values_list('dst_node', flat=True)
         root_nodes_id = AbstractNode.objects.filter(id__in=node_id_list).values_list('root', flat=True).distinct()
-        stuck_regs = AbstractNode.objects.filter(id__in=root_nodes_id, is_deleted=False)
+        stuck_regs = AbstractNode.objects.filter(id__in=root_nodes_id)
```
Comment (mfraezz) [inline osf/models/registrations.py:137]: This should include `is_deleted=False`
